src/main/java/com/limechain/runtime/Runtime.java

```diff
@@ -63,87 +72,40 @@ private byte[] callInner(String functionName, RuntimePointerSize parameterPtrSiz
         }
 
         RuntimePointerSize responsePtrSize = new RuntimePointerSize((long) response[0]);
-        return getDataFromMemory(responsePtrSize);
-    }
-
-    /**
-     * This setter exists to be used only in the RuntimeBuilder, since the RuntimeVersion is essentially an attribute
-     * of the Runtime, thus modeled as its field. If we can't find it in a custom section directly from the binary
-     * though, we'll still need an instance of `Runtime` in order to obtain the version by calling `Core_version`,
-     * and set it afterward via this setter.
-     */
-    void setVersion(RuntimeVersion runtimeVersion) {
-        this.version = runtimeVersion;
-    }
-
-    public int getHeapBase() {
-        return instance.exports.getGlobal(WasmExports.HEAP_BASE.getValue()).getIntValue();
-    }
-
-    public int getDataEnd() {
-        return instance.exports.getGlobal(WasmExports.DATA_END.getValue()).getIntValue();
-    }
-
-    private Memory getMemory() {
-        return instance.exports.getMemory(WasmExports.MEMORY.getValue());
+        return context.getSharedMemory().readData(responsePtrSize);
     }
 
     /**
-     * Get the data stored in memory using a {@link  RuntimePointerSize}
-     *
-     * @param runtimePointerSize pointer to data and its size
-     * @return byte array with read data
+     * @return the {@link RuntimeVersion} of this runtime instance.
+     * @implNote The runtime version is cached in the context, so no actual runtime invocation is performed.
      */
-    public byte[] getDataFromMemory(RuntimePointerSize runtimePointerSize) {
-        ByteBuffer memoryBuffer = this.getMemory().buffer();
-        byte[] data = new byte[runtimePointerSize.size()];
-        memoryBuffer.position(runtimePointerSize.pointer());
-        memoryBuffer.get(data);
-        return data;
+    public RuntimeVersion getVersion() {
+        return this.context.getRuntimeVersion();
     }
 
-    /**
-     * Write data to memory, by allocating space in memory and then writing to it.
-     *
-     * @param data data to be written
-     * @return a pointer size to the written data
-     */
-    public RuntimePointerSize writeDataToMemory(byte[] data) {
-        RuntimePointerSize allocatedPointer = allocate(data.length);
-        writeDataToMemory(data, allocatedPointer);
-        return allocatedPointer;
+    RuntimeVersion callCoreVersion() {
+        return ScaleUtils.Decode.decode(this.call("Core_version"), new RuntimeVersionReader());
```

Comment:
I feel like "Core_version" "Core_execute_block" and the rest should be listed as constants somewhere.

Reply:
I feel the same. Didn't know how far to push it, but I'd also introduce some kind of modelling for the Runtime API.

Reply:
Not sure what that modelling would look like so you might stick to the few exported constants for a start.

Reply:
That's exactly what I meant - constants for the runtime api endoints and methods in `Runtime` for calling them. For now, an `executeBlock` makes sense. Delegated to the follow-up PR.

Reply:
Proposed fix in [e222d87](https://github.com/LimeChain/Fruzhin/pull/464/commits/e222d870a46afcb797ef1ab9b9517f2b8c8bd6dc)

You might wanna take a look and see if it's fine for now.